07-errors.md

```diff
@@ -0,0 +1,284 @@
+| 18    | `ErrForbidden`         | operation forbidden                                 |
+| 19    | `ErrInsufficientFunds` | insufficient funds for operation                    |
+| 20    | `ErrIllegalState`      | the actor is in an illegal state                    |
+| 21    | `ErrSerialization`     | actor failed to serialize or deserialize some state |
+| 22    | `ErrInvalidMethod`     | specified method does not exist                     |
+| 23    | `ErrUnspecified`       | the actor failed with an unspecified error          |
+| 24-31 | reserved               |                                                     |
+
+Changes from pre-FVM Filecoin:
+
+1. Added `ErrInvalidMethod` (replacing `SysErrInvalidMethod`).
+2. Added `ErrUnspecified`. This generally means "there's a bug and I have no idea what went wrong".
+   The alternative would be for the actor to _panic_ (leading to an `ErrIllegalActor` syscall
+   error), but it may be useful to distinguish between "crashing" and "something unknown went
+   wrong".
+
+## Syscall Errors
+
+Syscall error codes are returned by syscalls to actors. Except `Ok` (0), they indicate that the
+syscall failed (without any side effects). The actor may choose to handle the error and continue, or
+abort (usually with a standard "abnormal exit" code).
+
+| Code | Status                 | Description                                            |
+|------|------------------------|--------------------------------------------------------|
+| 0    | `Ok`                   | syscall succeeded                                      |
+| 1    | `ErrIllegalArgument`   | invalid syscall parameters                             |
+| 2    | `ErrIllegalActor`      | actor is not in the correct state to perform operation |
+| 3    | `ErrLimit`             | some limit was exceeded (e.g. lookback limit)          |
+| 4    | `ErrAssertionFailed`   | some internal assertion failed                         |
+| 5    | `ErrInsufficientFunds` | attempted to send more than balance                    |
+| 6    | `ErrNotFound`          | resource not found                                     |
+| 7    | `ErrInvalidHandle`     | block handle invalid                                   |
+| 8    | `ErrIllegalCid`        | cid creation parameters (hash/length) were invalid     |
+| 9    | `ErrIllegalCodec`      | specified codec is not allowed                         |
+| 10   | `ErrSerialization`     | block format did not match specified codec             |
+
+**Notes:**
+
+- We intentionally use `ErrIllegalArgument` instead of `ErrSerialization` in non-IPLD syscalls, even
+    if we're using CBOR.
+- There is no `ErrForbidden` syscall error code:
+    - Calling a forbidden actor will result in an `ErrForbidden` exit code from that actor.
+    - Calling a forbidden syscall is statically impossible as it will be impossible to _import_ said
+        syscalls. I.e., actor deployment will fail.
```

Comment:
More of a strategy question for my own understanding, is the idea to never revoke syscalls? This doesn't always bear out in practice, but I suppose that this could be added if and when that happens.

Reply:
Yeah, you're right. And we may also come across cases where a _syscall_ isn't restricted, but some _parameter_ is restricted.